[DNM] SIMD

stephentyrone opened the pull request:
Initial implementation of simd types and protocols for the standard library. Pending Swift-Evolution review.

stephentyrone committed:
WIP

stephentyrone committed:
WIP.

stephentyrone committed:
Rework in progress.

stephentyrone committed:
Further work on SIMD, getting ready for pitch.

stephentyrone committed:
WIP

stephentyrone committed:
Further WIP.

stephentyrone committed:
Remove out-of-date floatingpoint simd work, add some properties.

stephentyrone committed:
Some additional comments.

stephentyrone committed:
Added FloatingPoint vector files.

stephentyrone committed:
Added floating-point vector ops for rounding.

stephentyrone committed:
Implement random( ) for vector types.

stephentyrone committed:
Added some additional conversions, reworked bitPattern for FloatingPoint.

stephentyrone committed:
get integer types the same way we do for scalar ints.

stephentyrone committed:
WIP.

stephentyrone committed:
Rename SIMDPredicateMxN -> SIMD.MaskM.VectorN

stephentyrone committed:
Finished removing Collection conformance.

stephentyrone commented:
@swift-ci please test

xedin commented:
@stephentyrone Quick question - would new types be loaded together with existing integer types or would they require additional `import`s? I'm asking because I see that they add a number of new operator overloads, so we'd need to consider what type-checker impact is going to be...

rudkx commented:
@stephentyrone If `SIMDVector` is `ExpressibleByArrayLiteral` and `SIMDFloatingPointVector` provides `func +`, how is one (person or compiler) supposed to interpret `[1,1,1,1] + [1,1,1,1]`?

stephentyrone commented:
@rudkx If I had my choice, we wouldn't interpret it, because using `+` for concatenation is a damned abomination whether or not literals are involved 😂.

... but that ship has sailed, and as a pragmatic matter I would prefer to interpret it as [Int] + [Int], since that's preserves the existing behavior, but that's not what currently falls out.

stephentyrone committed:
fromArray -> _

stephentyrone committed:
You get a dot, and YOU get a dot, and YOU get a dot!

Experimentally dot-prefix every elementwise simd operation. YOLO!

stephentyrone committed:
Ok, you don't *all* get dots after all.

Scale back .-prefixing operators to only the cases where it resolves
ambiguity between elementwise and structure operators.

(==, !=, *, &*).

stephentyrone committed:
Switch to using 'normal' logical operators.

moiseev commented on stdlib/public/core/SIMDInteger.swift:
```diff
@@ -0,0 +1,484 @@
+  @inlinable
+  static func >= <Scalar>(lhs: Scalar, rhs: Self) -> Mask
+  where Scalar : BinaryInteger {
+    return rhs <= rhs
```
One of the `rhs` should be `lhs`.

moiseev replied:
```suggestion
    return rhs <= lhs
```

moiseev committed:
Dot-prefixed operators

moiseev committed:
Fix the typo in SIMDIntegerVector.<=

stephentyrone committed:
Straighten out .-prefixing per core team discussion.

moiseev committed:
Allow vector types wider than 512 bits

moiseev committed:
Make stdlib compile in presence of all new operator overloads

moiseev committed:
Enable SIMD.VectorN<T> syntax for vector types

This commit adds a new SIMDVectorizable protocol that contains
associated types for all possible VectorN types, to which all integer
and floating point types trivially conform.

A set of new generic type aliases is also added to enable a more
idiomatic SIMD.Vector4<Int> syntax for the vector types.
